projects/optic/src/commands/diff/diff-all.ts

```diff
@@ -0,0 +1,365 @@
+      '--ruleset <ruleset>',
+      'run comparison with a locally defined ruleset, if not set, looks for the ruleset on the [x-optic-ruleset] key on the spec, and then the optic.dev.yml file.'
+    )
+    .option('--check', 'enable checks', false)
+    .option('--web', 'view the diff in the optic changelog web view', false)
+    .action(wrapActionHandlerWithSentry(getDiffAllAction(config)));
+};
+
+type DiffAllActionOptions = {
+  compareTo?: string;
+  compareFrom: string;
+  ruleset?: string;
+  check: boolean;
+  web: boolean;
+  json: boolean;
+};
+
+// Match up the to and from candidates
+// This will return the comparisons we can try to run
+function matchCandidates(
+  from: {
+    ref: string;
+    paths: string[];
+  },
+  to: { ref?: string; paths: string[] }
+): Map<
+  string,
+  {
+    from?: string;
+    to?: string;
+  }
+> {
+  const results = new Map<string, { from?: string; to?: string }>();
+  for (const path of from.paths) {
+    const strippedPath = path.replace(`${from.ref}:`, '');
+    results.set(strippedPath, {
+      from: path,
+    });
+  }
+
+  for (const path of to.paths) {
+    const strippedPath = to.ref ? path.replace(`${to.ref}:`, '') : path;
+    const maybePathObject = results.get(strippedPath);
+    if (maybePathObject) {
+      maybePathObject.to = path;
+    } else {
+      results.set(strippedPath, {
+        to: path,
+      });
+    }
+  }
+  return results;
+}
+
+async function computeAll(
+  candidatesMap: ReturnType<typeof matchCandidates>,
+  config: OpticCliConfig,
+  options: DiffAllActionOptions
+): Promise<{
+  warnings: Warnings;
+  results: Results;
+}> {
+  const warnings: Warnings = {
+    missingOpticUrl: [],
+    unparseableFromSpec: [],
+    unparseableToSpec: [],
+  };
+
+  const results: Results = [];
+
+  for await (const [_, candidate] of candidatesMap) {
+    // try load both from + to spec
+    let fromParseResults: ParseResult;
+    let toParseResults: ParseResult;
+    try {
+      fromParseResults = await getFileFromFsOrGit(
+        candidate.from,
+        config,
+        false
+      );
+    } catch (e) {
+      warnings.unparseableFromSpec.push({
+        path: candidate.from!,
+        error: e,
+      });
+      continue;
+    }
+
+    try {
+      toParseResults = await getFileFromFsOrGit(candidate.to, config, true);
+    } catch (e) {
+      warnings.unparseableToSpec.push({
+        path: candidate.to!,
+        error: e,
+      });
+      continue;
+    }
+
+    // Cases we run the comparison:
+    // - if to spec has x-optic-url
+    // - if from spec has x-optic-url AND to spec is empty
+    if (
+      typeof toParseResults.jsonLike[OPTIC_URL_KEY] === 'string' ||
+      (typeof fromParseResults.jsonLike[OPTIC_URL_KEY] === 'string' &&
+        toParseResults.isEmptySpec)
+    ) {
+      const { specResults, checks } = await compute(
+        [fromParseResults, toParseResults],
+        config,
+        options
+      );
+
+      if (config.isAuthenticated) {
```

Comment:
This feels a little weird to be here (this will mean we run the diff and upload and open in web, before moving on to the next command).

The reason it's here is because `from`, `to` and `results` can take up a lot of memory, and sequentially processing it means we don't need to load all results into memory. 

That being said, there's a couple of things:
- bulk-compare (from optic-ci) loads all comparisons into memory, and then dumps out results / uploads and works for our current usage
- It might be less of an issue when we optimize the changelog / results output

I think it's probably _fine_ here, but it can lead to _some_ partial runs if there's exceptions thrown